File: frontend/lib/api/stats.ts
import { apiGet, apiPost } from "./client";

export type PopularPost = {
  id: number;
  title: string;
  slug: string;
  views: number;
};

export type PostStats = {
  id: number;
  title: string;
  slug: string;
  views: number;
  unique: number;
  last_view_at: string | null;
};

export type DailyStats = {
  date: string;
  views: number;
  unique: number;
};

export type StatsSummary = {
  total_views: number;
  total_unique: number;
  views_today: number;
  unique_today: number;
};

export type AdminStats = {
  summary: StatsSummary;
  posts: PostStats[];
  daily: DailyStats[];
};

/** Зарегистрировать просмотр поста. Возвращает текущее число просмотров. */
export async function recordPostView(slug: string): Promise<number> {
  const res = await apiPost<{ views: number }>(`/api/v1/posts/${slug}/view`, {});
  return res.views;
}

/** Получить число просмотров поста */
export async function getPostViews(slug: string): Promise<number> {
  const res = await apiGet<{ views: number }>(`/api/v1/posts/${slug}/views`);
  return res.views;
}

/** Топ-N популярных постов */
export async function getPopularPosts(limit = 5): Promise<PopularPost[]> {
  const res = await apiGet<{ items: PopularPost[] }>(
    `/api/v1/posts/popular?limit=${limit}`
  );
  return res.items ?? [];
}

/** Полная статистика для админки */
export async function getAdminStats(days = 30): Promise<AdminStats> {
  return apiGet<AdminStats>(`/api/v1/admin/stats?days=${days}`);
}

File: backend/internal/model/stats.go
package model

import "time"

// PopularPost пост в блоке «Популярное»
type PopularPost struct {
	ID    int64  `json:"id"`
	Title string `json:"title"`
	Slug  string `json:"slug"`
	Views int64  `json:"views"`
}

// PostStats статистика поста для админки
type PostStats struct {
	ID         int64      `json:"id"`
	Title      string     `json:"title"`
	Slug       string     `json:"slug"`
	Views      int64      `json:"views"`
	Unique     int64      `json:"unique"`
	LastViewAt *time.Time `json:"last_view_at,omitempty"`
}

// DailyStats просмотры за день
type DailyStats struct {
	Date   string `json:"date"`
	Views  int64  `json:"views"`
	Unique int64  `json:"unique"`
}

// StatsSummary сводка по сайту
type StatsSummary struct {
	TotalViews  int64 `json:"total_views"`
	TotalUnique int64 `json:"total_unique"`
	ViewsToday  int64 `json:"views_today"`
	UniqueToday int64 `json:"unique_today"`
}

// AdminStats полный ответ статистики для админки
type AdminStats struct {
	Summary StatsSummary `json:"summary"`
	Posts   []PostStats  `json:"posts"`
	Daily   []DailyStats `json:"daily"`
}
